public/app/features/alerting/AlertTabCtrl.ts

```diff
@@ -272,20 +273,46 @@ export class AlertTabCtrl {
           foundTarget = firstTarget;
         } else {
           this.error = 'Could not find any metric queries';
+          return;
         }
       }
 
       const datasourceName = foundTarget.datasource || this.panel.datasource;
-      this.datasourceSrv.get(datasourceName).then(ds => {
-        if (!ds.meta.alerting) {
-          this.error = 'The datasource does not support alerting queries';
-        } else if (ds.targetContainsTemplate && ds.targetContainsTemplate(foundTarget)) {
-          this.error = 'Template variables are not supported in alert queries';
-        } else {
-          this.error = '';
-        }
-      });
+      promises.push(
+        this.datasourceSrv
+          .get(datasourceName)
+          .then(
+            (foundTarget => (ds: DataSourceApi) => {
+              if (!ds.meta.alerting) {
+                return 'The datasource does not support alerting queries';
+              } else if (ds.targetContainsTemplate && ds.targetContainsTemplate(foundTarget)) {
+                return 'Template variables are not supported in alert queries';
+              }
+              return '';
+            })(foundTarget)
+          )
+          .then(
+            err =>
+              new Promise((resolve, reject) => {
+                if (err) {
+                  reject(err);
+                } else {
+                  resolve();
+                }
```

Comment:
this looks pretty pointless, why not just throw error in the callback above?

Reply:
i rewrite it by $q